pandas/core/missing.py

```diff
@@ -182,14 +192,40 @@ def interpolate_1d(xvalues, yvalues, method='linear', limit=None,
     # contain indices of NaNs at the beginning of the series, and NaNs that
     # are more than'limit' away from the prior non-NaN.
 
+    # In case that maxgap is provided, preserve_nans is derived so that
+    # gaps with continuous NaN values of width > maxgap will be preserved.
+
     # set preserve_nans based on direction using _interp_limit
-    if limit_direction == 'forward':
-        preserve_nans = start_nans | set(_interp_limit(invalid, limit, 0))
-    elif limit_direction == 'backward':
-        preserve_nans = end_nans | set(_interp_limit(invalid, 0, limit))
+    if maxgap is None:
+        if limit_direction == 'forward':
```

Comment:
can you make this into another function that you call, this is already getting to be a huge function.